### PR TITLE
Push NATS error messages to error queue

### PR DESCRIPTION
**Pull request by Amusement**
Addressed comments -- removed max retries cap, improved error handling.

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/724)

### CODE REVIEW

File: mqtrigger/messageQueue/nats.go [hunk 1]
```diff
@@ -107,35 +107,68 @@ func msgHandler(nats *Nats, trigger *crd.MessageQueueTrigger) func(*ns.Msg) {
...
 		for k, v := range headers {
 			req.Header.Add(k, v)
 		}
 
-		// Make the request
-		resp, err := http.DefaultClient.Do(req)
-		if err != nil {
-			log.Warningf("Request failed: %v", url)
+		var resp *http.Response
+		for attempt := -1; attempt < trigger.Spec.MaxRetries; attempt++ {
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: This is entirely subjective but I'd have this loop go `attempt := 0; attempt <= trigger.Spec.MaxRetries; ...`

File: mqtrigger/messageQueue/nats.go [hunk 2]
```diff
@@ -107,35 +107,68 @@ func msgHandler(nats *Nats, trigger *crd.MessageQueueTrigger) func(*ns.Msg) {
...
+			// Make the request
+			log.Warningf("Request : %v", req)
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: remove

File: mqtrigger/messageQueue/nats.go [hunk 3]
```diff
@@ -107,35 +107,68 @@ func msgHandler(nats *Nats, trigger *crd.MessageQueueTrigger) func(*ns.Msg) {
...
+				continue
+			}
+			if err == nil && resp.StatusCode == 200 {
+				// Success, quit retries
+				break
+			}
+		}
+
+		if resp == nil {
+			log.Warning("The response was undefined. Quit.")
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: This means every retry failed, right?  That's probably a better log message.

File: mqtrigger/messageQueue/nats.go [hunk 4]
```diff
@@ -107,35 +107,68 @@ func msgHandler(nats *Nats, trigger *crd.MessageQueueTrigger) func(*ns.Msg) {
...
 			return
 		}
+
+		// TODO : Sequence
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: Hm?

File: mqtrigger/messageQueue/nats.go [hunk 5]
```diff
@@ -107,35 +107,68 @@ func msgHandler(nats *Nats, trigger *crd.MessageQueueTrigger) func(*ns.Msg) {
...
+		body, bodyErr := ioutil.ReadAll(resp.Body)
+		if bodyErr != nil {
+			log.Warningf("Response body error: %v", bodyErr)
 			return
 		}
-		if resp.StatusCode != 200 {
-			log.Printf("Request returned failure: %v", resp.StatusCode)
-			return
+
+		// TODO : what if err is not nil, then body is empty. publish either of them, not just body.
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: We shouldn't mix those errors into the same queue, simply avoiding publishing when body is empty is probably good enough.

File: mqtrigger/messageQueue/nats.go [hunk 6]
```diff
@@ -107,35 +107,68 @@ func msgHandler(nats *Nats, trigger *crd.MessageQueueTrigger) func(*ns.Msg) {
...
+
+		// Only the latest error response will be published to error topic
+		if err != nil || resp.StatusCode != 200 {
+			if len(trigger.Spec.ErrorTopic) > 0 {
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: `&& len(body) > 0`

File: mqtrigger/messageQueue/nats.go (hunk 5 above)
Comment (Amusement) [reply mqtrigger/messageQueue/nats.go]: This comment was actually added by @smruthi2187. I'll remove this TODO and check if the body is empty before publishing.

**Comment by life1347** [comment]
Reviewed 4 of 6 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[fission/main.go, line 197 at r2](https://reviewable.io/reviews/fission/fission/724#-LEeHBZL3KZRmCFsBjuM:-LEeHBZL3KZRmCFsBjuN:bop8jjv) ([raw file](https://github.com/fission/fission/blob/a252fb317a77c9ee8ca33f77902e0316e41c8c7b/fission/main.go#L197)):*
> ```Go
> 	mqtRespTopicFlag := cli.StringFlag{Name: "resptopic", Usage: "Topic that the function response is sent on (optional; response discarded if unspecified)"}
> 	mqtErrorTopicFlag := cli.StringFlag{Name: "errortopic", Usage: "Topic that the function error messages are sent to (optional; errors discarded if unspecified"}
> 	mqtMaxRetries := cli.IntFlag{Name: "maxretries", Usage: "Maximum number of times the function will be retried upon failure (optional; default is 0)"}
> ```

Use "Value" to set the default value 0 like
```
targetcpu := cli.IntFlag{Name: "targetcpu", Value: 80, Usage: "Target average CPU usage percentage across pods for scaling"}
```

---

*[fission/mqtrigger.go, line 163 at r2](https://reviewable.io/reviews/fission/fission/724#-LEeDS2K8ORTmgNQ0fFe:-LEeDS2K8ORTmgNQ0fFf:bc18ku9) ([raw file](https://github.com/fission/fission/blob/a252fb317a77c9ee8ca33f77902e0316e41c8c7b/fission/mqtrigger.go#L163)):*
> ```Go
> 	}
> 	// Default number of max retries is 0
> 	mqt.Spec.MaxRetries = 0
> ```

This line should be removed or it will overwrite previously maxRetries setting if maxRetries has been set.

---

*[mqtrigger/messageQueue/nats.go, line 133 at r2](https://reviewable.io/reviews/fission/fission/724#-LEeHluzBbwuuN-jYwJe:-LEeHluzBbwuuN-jYwJf:bctrzkz) ([raw file](https://github.com/fission/fission/blob/a252fb317a77c9ee8ca33f77902e0316e41c8c7b/mqtrigger/messageQueue/nats.go#L133)):*
> ```Go
> 			resp, err = http.DefaultClient.Do(req)
> 			if err != nil || resp == nil {
> 				// Retry without referencing status code of nil response on the next line
> ```

Print error message if err is not nil

---

*[mqtrigger/messageQueue/nats.go, line 139 at r2](https://reviewable.io/reviews/fission/fission/724#-LEeI9Ys7Y-kR7OOTJFL:-LEeI9Ys7Y-kR7OOTJFM:bsr4c6j) ([raw file](https://github.com/fission/fission/blob/a252fb317a77c9ee8ca33f77902e0316e41c8c7b/mqtrigger/messageQueue/nats.go#L139)):*
> ```Go
> 				// Success, quit retrying
> 				break
> 			}
> ```

Suggest using HTTP status const for the status code. (https://golang.org/pkg/net/http/#pkg-constants).

---


*Comments from [Reviewable](https://reviewable.io/reviews/fission/fission/724)*

**Comment by life1347** [comment]
Review status: 3 of 6 files reviewed, 11 unresolved discussions (waiting on @life1347, @soamvasani, and @Amusement)

---

*[mqtrigger/messageQueue/nats.go, line 133 at r3](https://reviewable.io/reviews/fission/fission/724#-LEnPMSc78u86ZMFqRP-:-LEnPMSc78u86ZMFqRP0:b2ila06) ([raw file](https://github.com/fission/fission/blob/0c2dbd091a96d7c975dbd04e1018ccc4c6b5db9a/mqtrigger/messageQueue/nats.go#L133)):*
> ```Go
> 			resp, err = http.DefaultClient.Do(req)
> 			if err != nil {
> 				log.Error(err)
> ```

We need to also print some useful information for debugging, how about 
```
log.Errorf("Error invoking function for trigger %v: %v", trigger.Metadata.Name, err)
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/fission/fission/724)*

**Comment by Amusement** [comment]
Review status: 2 of 6 files reviewed, 11 unresolved discussions (waiting on @life1347, @soamvasani, and @Amusement)

---

*[fission/mqtrigger.go, line 163 at r2](https://reviewable.io/reviews/fission/fission/724#-LEeDS2K8ORTmgNQ0fFe:-LFcUL621xtBDGdwG5ik:b-896fix) ([raw file](https://github.com/fission/fission/blob/a252fb317a77c9ee8ca33f77902e0316e41c8c7b/fission/mqtrigger.go#L163)):*
<details><summary><i>Previously, life1347 (Ta-Ching Chen) wrote…</i></summary><blockquote>

This line should be removed or it will overwrite previously maxRetries setting if maxRetries has been set.
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 129 at r1](https://reviewable.io/reviews/fission/fission/724#-LEGECmSl1G49p9Cagvl-r1-129:-LFcUWQG5F1xpy0-eIgs:b-896fix) ([raw file](https://github.com/fission/fission/blob/d7ec33e02cd98790d4774f81a875365b7fa97ac0/mqtrigger/messageQueue/nats.go#L129)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

This is entirely subjective but I'd have this loop go `attempt := 0; attempt <= trigger.Spec.MaxRetries; ...`
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 131 at r1](https://reviewable.io/reviews/fission/fission/724#-LEGECmSl1G49p9Cagvl-r1-131:-LFcUYGVEZbl4tkK0elZ:b-896fix) ([raw file](https://github.com/fission/fission/blob/d7ec33e02cd98790d4774f81a875365b7fa97ac0/mqtrigger/messageQueue/nats.go#L131)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 144 at r1](https://reviewable.io/reviews/fission/fission/724#-LEGECmSl1G49p9Cagvl-r1-144:-LFcUmeYDQvUmcj2_WLZ:b-896fix) ([raw file](https://github.com/fission/fission/blob/d7ec33e02cd98790d4774f81a875365b7fa97ac0/mqtrigger/messageQueue/nats.go#L144)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

This means every retry failed, right?  That's probably a better log message.
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 157 at r1](https://reviewable.io/reviews/fission/fission/724#-LEGECmSl1G49p9Cagvl-r1-157:-LFcUpldBx9YEz1-r2tv:b-896fix) ([raw file](https://github.com/fission/fission/blob/d7ec33e02cd98790d4774f81a875365b7fa97ac0/mqtrigger/messageQueue/nats.go#L157)):*
<details><summary><i>Previously, Amusement (Nafisa Shazia) wrote…</i></summary><blockquote>

This comment was actually added by @smruthi2187. I can remove this TODO and check if the body is empty before publishing. 
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 161 at r1](https://reviewable.io/reviews/fission/fission/724#-LEGECmSl1G49p9Cagvl-r1-161:-LFcUqa3EuX7KvS-7aCJ:b-896fix) ([raw file](https://github.com/fission/fission/blob/d7ec33e02cd98790d4774f81a875365b7fa97ac0/mqtrigger/messageQueue/nats.go#L161)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

`&& len(body) > 0`
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 133 at r2](https://reviewable.io/reviews/fission/fission/724#-LEeHluzBbwuuN-jYwJe:-LFcUk_CAHwhCnNqWeKL:b-896fix) ([raw file](https://github.com/fission/fission/blob/a252fb317a77c9ee8ca33f77902e0316e41c8c7b/mqtrigger/messageQueue/nats.go#L133)):*
<details><summary><i>Previously, life1347 (Ta-Ching Chen) wrote…</i></summary><blockquote>

Print error message if err is not nil
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 139 at r2](https://reviewable.io/reviews/fission/fission/724#-LEeI9Ys7Y-kR7OOTJFL:-LFcUlkZ0EEUwgKm6dQN:b-896fix) ([raw file](https://github.com/fission/fission/blob/a252fb317a77c9ee8ca33f77902e0316e41c8c7b/mqtrigger/messageQueue/nats.go#L139)):*
<details><summary><i>Previously, life1347 (Ta-Ching Chen) wrote…</i></summary><blockquote>

Suggest using HTTP status const for the status code. (https://golang.org/pkg/net/http/#pkg-constants).
</blockquote></details>

Done.

---

*[mqtrigger/messageQueue/nats.go, line 133 at r3](https://reviewable.io/reviews/fission/fission/724#-LEnPMSc78u86ZMFqRP-:-LFcUYnq208pMwseVgbf:b-896fix) ([raw file](https://github.com/fission/fission/blob/0c2dbd091a96d7c975dbd04e1018ccc4c6b5db9a/mqtrigger/messageQueue/nats.go#L133)):*
<details><summary><i>Previously, life1347 (Ta-Ching Chen) wrote…</i></summary><blockquote>

We need to also print some useful information for debugging, how about 
```
log.Errorf("Error invoking function for trigger %v: %v", trigger.Metadata.Name, err)
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/fission/fission/724)*

**Comment by life1347** [comment]
LGTM

**Pull request merged**